### ISSUE TITLE
Conda build fails with "undefined reference to `mlr_dsl_ParseTrace'"

### ISSUE THREAD

**Issue by BEFH** [johnkerl/miller#740]
When I compile the latest version (5.10.2) without using anaconda, it successfully compiles and runs the unit tests, but when I try and build it as an anaconda package, I get the following error while it runs `make`:

```
/bin/sh ../libtool  --tag=CC   --mode=link $BUILD_PREFIX/bin/x86_64-conda-linux-gnu-cc -Wall -std=gnu99 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/miller-5.10.2 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -static -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,$PREFIX/lib -Wl,-rpath-link,$PREFIX/lib -L$PREFIX/lib -o mlr mlrmain.o cli/libcli.la containers/libcontainers.la stream/libstream.la input/libinput.la dsl/libdsl.la mapping/libmapping.la output/liboutput.la lib/libmlr.la parsing/libdsl.la auxents/libauxents.la -lm
libtool: link: $BUILD_PREFIX/bin/x86_64-conda-linux-gnu-cc -Wall -std=gnu99 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/miller-5.10.2 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z -Wl,relro -Wl,-z -Wl,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath -Wl,$PREFIX/lib -Wl,-rpath-link -Wl,$PREFIX/lib -o mlr mlrmain.o  -L$PREFIX/lib cli/.libs/libcli.a containers/.libs/libcontainers.a stream/.libs/libstream.a input/.libs/libinput.a dsl/.libs/libdsl.a mapping/.libs/libmapping.a output/.libs/liboutput.a lib/.libs/libmlr.a parsing/.libs/libdsl.a auxents/.libs/libauxents.a -lm
/sc/arion/work/fultob01/conda/envs/py3.9/conda-bld/miller_1636495473849/_build_env/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: parsing/.libs/libdsl.a(mlr_dsl_wrapper.o): in function `mlr_dsl_parse':
mlr_dsl_wrapper.c:(.text.mlr_dsl_parse+0x103): undefined reference to `mlr_dsl_ParseTrace'
```

My build script is like so:

```sh
#!/bin/sh

./configure --prefix=$PREFIX
make
make check
make install
```

I have made the gcc toolchain, make and flex available. It also fails when I install gcc, make and flex using Anaconda then make manually.

Do you have any idea what's going on?

**Comment by BEFH** [comment]
The whole output is here: https://pastebin.com/8qNJ0R2j

**Comment by johnkerl** [comment]
> Do you have any idea what's going on?

kinda ... the `mlr_dsl_ParseTrace` has to do with an output of Lemon.

that said, Lemon (and all the Miller C code) are gone as of the "red PR" of https://github.com/johnkerl/miller/pull/731 so it's a bit of a moot point

more than anything, i would relieve you of any further work trying to find/fix the compile error :) -- miller 6 (soon to be released) is the forward path and any issues that arise will be different issues ...

**Comment by BEFH** [comment]
Do you mean https://lemon.cs.elte.hu/trac/lemon? Are the headers in your repo, or do they need to be on the `$LD_LIBRARY_PATH`?

It would be nice to fix this both for the go version and the soon-to-be-deprecated C version, just in case my group needs the old version for some reason.

**Comment by johnkerl** [comment]
Lemon was copied inside the Miller repo ... no `$LD_LIBRARY_PATH`. Lemon and all the C files aren't in the Miller repo anymore; if your team needs the C version, maybe download the 5.10.2 release and build from source with `./configure && make` -- ?

**Comment by BEFH** [comment]
That's basically what happens. It downloads https://github.com/johnkerl/miller/archive/refs/tags/v5.10.2.tar.gz, verifies the checksum, extracts the archive, descends into the directory, then runs the following:

```
./configure --prefix=$PREFIX
make
make check
make install
```

The dependencies in the meta.yaml file, including GCC, make and flex are installed first. If I do all of that manually using Anaconda instead of `conda-build`, it still fails, but if I install GCC and make using `apt`, it works. So it's likely something about the Anaconda version of GCC and make, or some dependency that's installed by the apt packages but not Anaconda.

**Comment by johnkerl** [comment]
at line 1025 of https://pastebin.com/8qNJ0R2j is the `sed` line which produces `mlr_dsl_ParseTrace` -- where does the _output_ of these commands go? if we could get the output of `grep ParseTrace mlr_dsl_parse.c.tmp` that might help debug ...

**Comment by BEFH** [comment]
OK, so I am currently going insane. I just made a new VM and installed using conda, and it didn't error, but when I did conda-build, still the same error.

`mlr_dsl_parse.c.tmp` has been deleted by the time the error happened, but the sed had run, I think:

```
mv mlr_dsl_parse.c mlr_dsl_parse.c.tmp
sed \
		-e 's/ParseTrace/mlr_dsl_ParseTrace/g' \
		-e 's/ParseTokenName/mlr_dsl_ParseTokenName/g' \
		-e 's/lemon_parser_alloc/mlr_dsl_lemon_parser_alloc/g' \
		-e 's/lemon_parser_free/mlr_dsl_lemon_parser_free/g' \
		-e 's/lemon_parser_parse_token/mlr_dsl_lemon_parser_parse_token/g' \
		-e 's/yy_destructor/mlr_dsl_yy_destructor/g' \
	mlr_dsl_parse.c.tmp > mlr_dsl_parse.c
rm -f mlr_dsl_parse.c.tmp
```

So I ran the following:

```
(base) ubuntu@charmed-dobsonfly:~/miniconda3/conda-bld/miller_1636515952848$ ll
total 28
drwxrwxr-x  6 ubuntu ubuntu 4096 Nov  9 22:46 ./
drwxrwxr-x  9 ubuntu ubuntu 4096 Nov  9 22:46 ../
drwxrwxr-x 12 ubuntu ubuntu 4096 Nov  9 22:46 _build_env/
drwxrwxr-x  7 ubuntu ubuntu 4096 Nov  9 22:46 _h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/
drwxrwxr-x  2 ubuntu ubuntu 4096 Nov  9 22:46 pip_cache/
-rw-rw-r--  1 ubuntu ubuntu 2307 Nov  9 22:46 prefix_files.txt
drwxrwxr-x 11 ubuntu ubuntu 4096 Nov  9 22:46 work/
(base) ubuntu@charmed-dobsonfly:~/miniconda3/conda-bld/miller_1636515952848$ find . -name mlr_dsl_parse.c
./work/c/parsing/mlr_dsl_parse.c
(base) ubuntu@charmed-dobsonfly:~/miniconda3/conda-bld/miller_1636515952848$ find . -name mlr_dsl_parse.c
./work/c/parsing/mlr_dsl_parse.c
(base) ubuntu@charmed-dobsonfly:~/miniconda3/conda-bld/miller_1636515952848$ grep ParseTrace work/c/parsing/mlr_dsl_parse.c
void mlr_dsl_ParseTrace(FILE *TraceFILE, char *zTracePrompt) {
```

The sed command appears to have worked.

**Comment by BEFH** [comment]
Here are the working and non-working environments:

Working:

```
  - _go_select=2.3.0=cgo
  - _libgcc_mutex=0.1=conda_forge
  - _openmp_mutex=4.5=1_llvm
  - binutils=2.36.1=hdd6e379_2
  - binutils_impl_linux-64=2.36.1=h193b22a_2
  - binutils_linux-64=2.36=hf3e587d_1
  - brotlipy=0.7.0=py39h27cfd23_1003
  - bzip2=1.0.8=h7f98852_4
  - c-ares=1.18.1=h7f98852_0
  - c-compiler=1.3.0=h7f98852_0
  - ca-certificates=2021.10.8=ha878542_0
  - certifi=2021.10.8=py39hf3d152e_1
  - cffi=1.14.6=py39h400218f_0
  - chardet=4.0.0=py39h06a4308_1003
  - conda=4.10.3=py39hf3d152e_3
  - conda-package-handling=1.7.3=py39h27cfd23_1
  - cryptography=3.4.7=py39hd23ed53_0
  - cxx-compiler=1.3.0=h4bd325d_0
  - flex=2.6.4=h58526e2_1004
  - gcc=9.4.0=h192d537_1
  - gcc_impl_linux-64=9.4.0=h03d3576_11
  - gcc_linux-64=9.4.0=h391b98a_1
  - go=1.16.9=h17d2266_0
  - gxx=9.4.0=h192d537_1
  - gxx_impl_linux-64=9.4.0=h03d3576_11
  - gxx_linux-64=9.4.0=h0316aca_1
  - icu=68.2=h9c3ff4c_0
  - idna=2.10=pyhd3eb1b0_0
  - kernel-headers_linux-64=2.6.32=he073ed8_15
  - krb5=1.19.2=hcc1bbae_0
  - ld_impl_linux-64=2.36.1=hea4e1c9_2
  - libarchive=3.5.2=hccf745f_1
  - libcurl=7.79.1=h2574ce0_1
  - libedit=3.1.20191231=he28a2e2_2
  - libev=4.33=h516909a_1
  - libffi=3.3=he6710b0_2
  - libgcc-devel_linux-64=9.4.0=hd854feb_11
  - libgcc-ng=11.2.0=h1d223b6_11
  - libgfortran-ng=11.2.0=h69a702a_11
  - libgfortran5=11.2.0=h5c6108e_11
  - libgomp=11.2.0=h1d223b6_11
  - libiconv=1.16=h516909a_0
  - libnghttp2=1.43.0=h812cca2_1
  - libsanitizer=9.4.0=h79bfe98_11
  - libsolv=0.7.19=h780b84a_5
  - libssh2=1.10.0=ha56f1ee_2
  - libstdcxx-devel_linux-64=9.4.0=hd854feb_11
  - libstdcxx-ng=11.2.0=he4da1e4_11
  - libxml2=2.9.12=h72842e0_0
  - llvm-openmp=12.0.1=h4bd325d_1
  - lz4-c=1.9.3=h9c3ff4c_1
  - lzo=2.10=h516909a_1000
  - m4=1.4.18=h516909a_1001
  - make=4.3=hd18ef5c_1
  - mamba=0.17.0=py39h951de11_0
  - ncurses=6.2=he6710b0_1
  - openssl=1.1.1l=h7f98852_0
  - pip=21.1.3=py39h06a4308_0
  - pycosat=0.6.3=py39h27cfd23_0
  - pycparser=2.20=py_2
  - pyopenssl=20.0.1=pyhd3eb1b0_1
  - pysocks=1.7.1=py39h06a4308_0
  - python=3.9.5=h12debd9_4
  - python_abi=3.9=2_cp39
  - readline=8.1=h27cfd23_0
  - reproc=14.2.3=h7f98852_0
  - reproc-cpp=14.2.3=h9c3ff4c_0
  - requests=2.25.1=pyhd3eb1b0_0
  - ruamel_yaml=0.15.100=py39h27cfd23_0
  - setuptools=52.0.0=py39h06a4308_0
  - six=1.16.0=pyhd3eb1b0_0
  - sqlite=3.36.0=hc218d9a_0
  - sysroot_linux-64=2.12=he073ed8_15
  - tk=8.6.10=hbc83047_0
  - tqdm=4.61.2=pyhd3eb1b0_1
  - tzdata=2021a=h52ac0ba_0
  - urllib3=1.26.6=pyhd3eb1b0_1
  - wheel=0.36.2=pyhd3eb1b0_0
  - xz=5.2.5=h7b6447c_0
  - yaml=0.2.5=h7b6447c_0
  - zlib=1.2.11=h7b6447c_3
  - zstd=1.5.0=ha95c52a_0
```

nonworking:

```
  - _libgcc_mutex=0.1=main
  - _openmp_mutex=4.5=1_gnu
  - _sysroot_linux-64_curr_repodata_hack=3=haa98f57_10
  - binutils_impl_linux-64=2.35.1=h27ae35d_9
  - binutils_linux-64=2.35.1=h454624a_30
  - gcc_impl_linux-64=9.3.0=h6df7d76_17
  - gcc_linux-64=9.3.0=h1ee779e_30
  - go-core=1.11.4=hf39692e_0
  - go_linux-64=1.11.4=h0f5337a_0
  - gxx_impl_linux-64=9.3.0=hbdd7822_17
  - gxx_linux-64=9.3.0=h7e70986_30
  - kernel-headers_linux-64=3.10.0=h57e8cba_10
  - ld_impl_linux-64=2.35.1=h7274673_9
  - libgcc-devel_linux-64=9.3.0=hb95220a_17
  - libgcc-ng=9.3.0=h5101ec6_17
  - libgfortran-ng=7.5.0=ha8ba4b0_17
  - libgfortran4=7.5.0=ha8ba4b0_17
  - libgomp=9.3.0=h5101ec6_17
  - libstdcxx-devel_linux-64=9.3.0=hf0c5c8d_17
  - libstdcxx-ng=9.3.0=hd4cf53a_17
  - make=4.2.1=h1bed415_1
  - sysroot_linux-64=2.17=h57e8cba_10
```

There are a lot of differences, and I have no idea what it is.

**Comment by johnkerl** [comment]
bummer

well the sed has to have 'worked' in the sense of finding its input & output files, else it would have errored out ... i think for it to have 'not worked' it would have to have not found any `ParseTrace` lines to replace with `mlr_dsl_ParseTrace` i.e. it simply didn't find the search pattern. hence my ask about doing a grep.

but now it's failing to fail. :(

**Comment by johnkerl** [comment]
@BEFH I've used conda exactly zero times but to be more help to you I should try tomorrow evening to reproduce what you're doing here ...

**Comment by BEFH** [comment]
Yeah, I tried searching for the other substitutions, and they seem to have worked. Let me put together a list of commands I used when it worked and when it didn't.

**Comment by johnkerl** [comment]
the only three reasons i can think of for `undefined reference to 'mlr_dsl_ParseTrace'` are (1) the sed not finding the search pattern; (2) a source file not being included in a particular compile; (3) C vs C++ name-mangling. #1 we ruled out of course, & i doubt #3 since there would be many more errors than just one ... re #2 from the pastebin output (line 1070-1080 region) i don't see the `libdsl.a` being omitted so i will definitely need to reproduce this locally ...

**Comment by BEFH** [comment]
Well, I've been unable to replicate my successful run, but this should get you the failure:

Always answer "yes" when prompted in the installation.

```
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash Miniconda3-latest-Linux-x86_64.sh
. .bashrc
conda install -y -c conda-forge c-compiler make flex
wget https://github.com/johnkerl/miller/archive/refs/tags/v5.10.2.tar.gz
tar -xf v5.10.2.tar.gz
cd miller-5.10.2
mkdir sandbox
./configure --prefix=$PWD/sandbox
make
```

I'm using `multipass` to make clean ubuntu VMs for testing:

```
multipass launch -m 4G -d 10G
multipass shell [VM NAME]
```

You can get the same thing to happen by cloning my feedstock repo, installing conda-build (`conda install -c conda-forge conda-build`), descending into the feedstock, and running `conda-build recipe/`, but the build environment and files are in obscure directories.

**Comment by johnkerl** [comment]
Hmm @BEFH the first method (`conda install -y -c ....`) resulted in a `mlr` executable -- it failed to fail. :(

The second method: I've done `conda install -c conda-forge conda-build` and it says it's installed, but I'm getting `conda-build: command not found` :(

**Comment by johnkerl** [comment]
@BEFH still debugging, haven't forgotten (nor succeeded) ;)

**Comment by johnkerl** [comment]
@BEFH after some time debugging the conda build & comparing various debug/print statements against the non-conda build, I must say I truly do not know what's going, why the build failure only inside conda.

Last night I thought I was onto something. Namely:

* This is happening in regenerating `.h` and `.c` files from the grammar-spec `.l` and `.y` files
* Those latter are not changing, so the parser-generation is, best-case, only re-creating `mlr_dsl_parse.h`, `mlr_dsl_parse.c`, et al. which are already in source control. In fact it was always unnecessary in some sense to build Lemon and run the parser-generation on installs -- this belongs more properly to be done at development/package-creation time (which is the strategy I'm pursuing for the Go port / Miller 6).
* So the spot that's breaking at install time is a step that does not need to be run at all at install time.

So ... I made a `https://johnkerl.org/miller-releases/miller-5.10.2.conda.tar.gz` which can be used in `recipe/meta.yaml`

The change is to replace https://github.com/johnkerl/miller/blob/v5.10.2/c/parsing/Makefile.am with just

```
noinst_LTLIBRARIES= libdsl.la
libdsl_la_SOURCES=  mlr_dsl_lexer.c mlr_dsl_lexer.h mlr_dsl_parse.c mlr_dsl_parse.h mlr_dsl_wrapper.c mlr_dsl_wrapper.h

AM_CPPFLAGS=  -I${srcdir}
```

But ... now it's just a straight-up linker error. The `mlr_dsl_parse.c` is the one from version control, and there's no longer a Lemon step re-generating the same file. Just a compile and ilnk, all `.h` and `.c` files from version control, which results in a symbol-not-found error.

So my next step is to do more side-by-side comparison of:

* On one side, `./configure` and `make` in a tarball-extract of `miller-5.10.2.conda`;
* On the other, `conda recipe build`

**Comment by johnkerl** [comment]
Also (all along) there was `libdsl.a` in `c/parsing` and `c/dsl`. I thought that that was something that maybe one toolchain would be OK with and another would not. However I tried renaming the one in `c/parsing` to `libparsing.a` (along with mods in all necessary `Makefile.am` files) but, alas, the same symbol-not-found error persists. :(

**Comment by johnkerl** [comment]
... LOL re above https://github.com/johnkerl/miller/issues/740#issuecomment-964770167 _the only three reasons ..._ -- i should have added a fourth, namely, preprocessor directives ;).

Inside `mlr_dsl_parse.c` is

```
#ifndef NDEBUG
...
void mlr_dsl_ParseTrace(FILE *TraceFILE, char *zTracePrompt) {
    yyTraceFILE = TraceFILE;
    yyTracePrompt = zTracePrompt;
    if (yyTraceFILE==0)
        yyTracePrompt = 0;
    else if (yyTracePrompt==0)
        yyTraceFILE = 0;
}
#endif /* NDEBUG */
```

and it turns out that the Conda build is building with `NDEBUG` defined. So `mlr_dsl_ParseTrace` is getting ifdeffed away. :(

I didn't see any references to `NDEBUG` in your `miller-feedstock` so perhaps I can further modify the `miller-5.10.2.conda` tarball to edit out `NDEBUG` references ...

**Comment by johnkerl** [comment]
... yes, this seems to work -- https://johnkerl.org/miller-releases/miller-5.10.2.conda.tar.gz you can try in your `recipe/meta/yaml` -- ?

**Comment by johnkerl** [comment]
@BEFH some options:

* If the Conda build can be tweaked to not compile with `-D NDEBUG` then there are no Miller source-code-level changes needing to be made
* If not, maybe you can point at https://johnkerl.org/miller-releases/miller-5.10.2.conda.tar.gz above
* If that doesn't work -- since https://github.com/BEFH/miller-feedstock/blob/master/recipe/meta.yaml#L11 seems to require that tarballs be hosted from https://github.com/johnkerl/miller/archive not https://johnkerl.org/miller-releases -- then I can cut a 5.10.3 release of Miller which is branched off of the last 5.10.2 commit (wherein the C code still existed)

**Comment by BEFH** [comment]
@johnkerl I can change the manifest file (and the checksum), but it will have to be changed back later and will not auto-update when the go version comes out. Let me test locally with that tarball, then if it works, you can decide what makes sense for when I submit the pull request to the conda-forge feedstock.

**Comment by BEFH** [comment]
@johnkerl  OK, so I tested with the local release you provided, and after a brief issue with me getting the checksum wrong, it now works perfectly and creates the binary and manfile. All I had to do was to change the checksum and download location.

But it seems to me that NDEBUG should NOT be enabled? It's not in my build.sh and it seems like it should only be there if the dev wants it there. Should we submit a bug report somewhere else?

**Comment by johnkerl** [comment]
> But it seems to me that NDEBUG should NOT be enabled? It's not in my build.sh and it seems like it should only be there if the dev wants it there. Should we submit a bug report somewhere else?

Gleah.

Welp, ideally I'd like it if people weren't (randomly, from my POV) adding `-D FOO` when my code uses `#ifndef FOO`. However ... this C code is all going away (gone, rather) -- and even if it weren't it'd be easier for me to change `NDEBUG` to `LEMON_NDEBUG` or `MLR_DSL_NDEBUG` or what have you within files I _do_ control.

So my $0.02 is, let's not open a bug report to Conda ...

**Comment by johnkerl** [comment]
@BEFH https://github.com/johnkerl/miller/releases/tag/v5.10.3 ... 🙏

**Comment by BEFH** [comment]
Sweet! Thanks for all your help on this. I've opened a pull request for the conda-forge feedstock, and all builds were successful for the CI.

https://github.com/conda-forge/miller-feedstock/pull/9

Now I'm just waiting on either one of the existing maintainers to approve it, or for me to be added as a maintainer (separate pull request).
